blockchain_api/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+    serde::Deserialize,
+    std::{collections::HashSet, convert::Infallible, sync::Arc, time::Duration},
+    tokio::{sync::RwLock, task::JoinHandle},
+    tracing::error,
+    url::Url,
+};
+
+const BLOCKCHAIN_API_SUPPORTED_CHAINS_ENDPOINT_STR: &str = "/v1/supported-chains";
+const BLOCKCHAIN_API_RPC_ENDPOINT_STR: &str = "/v1";
+const BLOCKCHAIN_API_RPC_CHAIN_ID_PARAM: &str = "chainId";
+const BLOCKCHAIN_API_RPC_PROJECT_ID_PARAM: &str = "projectId";
+
+const SUPPORTED_CHAINS_REFRESH_INTERVAL: Duration = Duration::from_secs(60 * 60 * 4);
+
+#[derive(Debug, Deserialize)]
+struct SupportedChainsResponse {
+    pub http: HashSet<String>,
+}
+
+#[derive(Debug, Clone)]
+pub struct BlockchainApiProvider {
+    project_id: ProjectId,
+    blockchain_api_rpc_endpoint: Url,
+    supported_chains: Arc<RwLock<HashSet<String>>>,
+    refresh_job: Arc<JoinHandle<Infallible>>,
+}
+
+impl Drop for BlockchainApiProvider {
+    fn drop(&mut self) {
+        self.refresh_job.abort();
+    }
+}
+
+async fn refresh_supported_chains(
+    blockchain_api_supported_chains_endpoint: Url,
+    supported_chains: &RwLock<HashSet<String>>,
+) -> Result<(), Error> {
+    let response = reqwest::get(blockchain_api_supported_chains_endpoint)
+        .await?
+        .json::<SupportedChainsResponse>()
+        .await?;
+    *supported_chains.write().await = response.http;
+    Ok(())
+}
+
+impl BlockchainApiProvider {
+    pub async fn new(project_id: ProjectId, blockchain_api_endpoint: Url) -> Result<Self, Error> {
+        let blockchain_api_rpc_endpoint = blockchain_api_endpoint
+            .join(BLOCKCHAIN_API_RPC_ENDPOINT_STR)
+            .expect("Safe unwrap: hardcoded URL: BLOCKCHAIN_API_RPC_ENDPOINT_STR");
+        let blockchain_api_supported_chains_endpoint = blockchain_api_endpoint
+            .join(BLOCKCHAIN_API_SUPPORTED_CHAINS_ENDPOINT_STR)
+            .expect("Safe unwrap: hardcoded URL: BLOCKCHAIN_API_SUPPORTED_CHAINS_ENDPOINT_STR");
+
+        let supported_chains = Arc::new(RwLock::new(HashSet::new()));
+        refresh_supported_chains(
+            blockchain_api_supported_chains_endpoint.clone(),
+            &supported_chains,
+        )
+        .await?;
+        let mut interval = tokio::time::interval(SUPPORTED_CHAINS_REFRESH_INTERVAL);
+        interval.tick().await;
+        let refresh_job = tokio::task::spawn({
+            let supported_chains = supported_chains.clone();
+            let blockchain_api_supported_chains_endpoint =
+                blockchain_api_supported_chains_endpoint.clone();
+            async move {
+                loop {
+                    interval.tick().await;
+                    if let Err(e) = refresh_supported_chains(
+                        blockchain_api_supported_chains_endpoint.clone(),
+                        &supported_chains,
+                    )
+                    .await
+                    {
+                        error!("Failed to refresh supported chains: {e}");
```

Comment:
Ideally we could get service-level metrics for this, but would take more refactoring and not a big deal w/o.